src/natcap/invest/validation.py

```diff
@@ -1007,55 +312,46 @@ def validate(args, spec, spatial_overlap_opts=None):
-        axis_keys = None
-        if param_type == 'csv':
-            axis_keys = ['columns', 'rows']
-        elif param_type == 'vector' or 'vector' in param_type:
-            axis_keys = ['fields']
-        elif param_type == 'directory':
-            axis_keys = ['contents']
+        axis_keys = set(dir(parameter_spec)).intersection({'columns', 'rows', 'fields', 'contents'})
 
         if axis_keys:
             for axis_key in axis_keys:
-                if axis_key not in parameter_spec:
+                if getattr(parameter_spec, axis_key) is None:
                     continue
-                for nested_key, nested_spec in parameter_spec[axis_key].items():
-                    if ('required' in nested_spec
-                            and isinstance(nested_spec['required'], str)):
-                        parameter_spec[axis_key][nested_key]['required'] = (
+                for nested_spec in getattr(parameter_spec, axis_key):
+                    if (isinstance(nested_spec.required, str)):
+                        nested_spec.required = (
                             bool(_evaluate_expression(
-                                nested_spec['required'], expression_values)))
-
-        type_validation_func = _VALIDATION_FUNCS[param_type]
-
-        if type_validation_func is None:
-            # Validation for 'other' type must be performed by the user.
-            continue
+                                nested_spec.required, expression_values)))
         try:
             # pass the entire arg spec into the validation function as kwargs
             # each type validation function allows extra kwargs with **kwargs
-            warning_msg = type_validation_func(args[key], **parameter_spec)
+            warning_msg = parameter_spec.validate(args[key])
             if warning_msg:
                 validation_warnings.append(([key], warning_msg))
                 invalid_keys.add(key)
         except Exception:
-            LOGGER.exception(
-                'Error when validating key %s with value %s',
-                key, args[key])
-            validation_warnings.append(([key], MESSAGES['UNEXPECTED_ERROR']))
+            LOGGER.exception(f'Error when validating key {key} with value {args[key]}')
+            validation_warnings.append(([key], get_message('UNEXPECTED_ERROR')))
 
     # Phase 3: Check spatial overlap if applicable
-    if spatial_overlap_opts:
-        spatial_keys = set(spatial_overlap_opts['spatial_keys'])
+    if model_spec.validate_spatial_overlap:
+
+        # validate_spatial_overlap can be True, meaning validate all spatial keys,
+        # or a list, representing a subset of keys to validate
+        if isinstance(model_spec.validate_spatial_overlap, list):
+            spatial_keys = set(model_spec.validate_spatial_overlap)
+        else:
+            spatial_keys = set()
+            for i in model_spec.inputs:
+                if i.type in['raster', 'vector']:
```

Comment:
@davemfish, here's an instance where `type` is used!

Reply:
Its main purpose is to indicate the input type when we JSON serialize it and send it to the workbench. (Maintaining the old API)